differs/gem_diff.go

```diff
@@ -0,0 +1,143 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package differs
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	pkgutil "github.com/GoogleCloudPlatform/container-diff/pkg/util"
+	"github.com/GoogleCloudPlatform/container-diff/util"
+	"github.com/sirupsen/logrus"
+)
+
+type GemAnalyzer struct {
+}
+
+func (g GemAnalyzer) Name() string {
+	return "GemAnalyzer"
+}
+
+// GemAnalyzer.Diff compares gem/bundler installed Ruby packages between layers of two different images
+func (g GemAnalyzer) Diff(image1, image2 pkgutil.Image) (util.Result, error) {
+	diff, err := multiVersionDiff(image1, image2, g)
+	return diff, err
+}
+
+func (g GemAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
+	analysis, err := multiVersionAnalysis(image, g)
+	return analysis, err
+}
+
+func (g GemAnalyzer) getPackages(image pkgutil.Image) (map[string]map[string]util.PackageInfo, error) {
+	path := image.FSPath
+	packages := make(map[string]map[string]util.PackageInfo)
+	rubyPaths := []string{}
+
+	if image.Config.Config.Env != nil {
+		paths := getGemPaths(image.Config.Config.Env)
+		for _, p := range paths {
+			rubyPaths = append(rubyPaths, p)
+		}
+	}
+
+	rubyVersions, err := getRubyVersion(path)
+
+	if err != nil {
+		// Image doesn't have Ruby installed
+		return packages, nil
+	}
+
+	// default UNIX ruby installation paths
+	for _, rubyVersion := range rubyVersions {
+		rubyPaths = append(rubyPaths, filepath.Join(path, "usr/lib/ruby/gems", rubyVersion))
+		rubyPaths = append(rubyPaths, filepath.Join(path, "usr/local/lib/ruby/gems", rubyVersion, "gems"))
+	}
+
+	for _, rubyPath := range rubyPaths {
+		contents, err := ioutil.ReadDir(rubyPath)
+		if err != nil {
+			// Ruby version folder doesn't have any gems
+			continue
+		}
+
+		for i := 0; i < len(contents); i++ {
+			c := contents[i]
+			fileName := c.Name()
+			packageDir := regexp.MustCompile("^([a-z|A-Z|0-9|_|-]+)-([0-9|\\.|a-z]+)$")
+			packageMatch := packageDir.FindStringSubmatch(fileName)
+			if len(packageMatch) != 0 {
+				packageName := packageMatch[1]
+				version := packageMatch[2][:len(packageMatch[2])]
+
+				// retrieves size of package
+				var size int64
+
+				if i-1 >= 0 {
+					packagePath := filepath.Join(rubyPath, packageName+"-"+version)
+					size = pkgutil.GetSize(packagePath)
+
+					currPackage := util.PackageInfo{Version: version, Size: size}
+					mapPath := strings.Replace(packagePath, path, "", 1)
+					addToMap(packages, packageName, mapPath, currPackage)
+				} else {
+					logrus.Errorf("Could not find RubyGem %s for corresponding metadata info", packageName)
+					continue
+				}
+			}
+		}
+	}
+
+	return packages, nil
+}
+
+func getRubyVersion(pathToLayer string) ([]string, error) {
+	matches := []string{}
+	libPath := filepath.Join(pathToLayer, "usr/local/lib/ruby/gems")
```

Comment:
I'm not sure you can rely on this always being the path to the ruby library? You might be able to use an env var like `RBENV_ROOT` to find it.